sklearn/utils/_random.pyx

```diff
@@ -303,6 +308,29 @@ cpdef sample_without_replacement(cnp.int_t n_population,
                          % (all_methods, method))
 
 
+def sample_without_replacement(
+        object n_population, object n_samples, method="auto", random_state=None):
+    cdef:
+        cnp.intp_t n_pop_intp, n_samples_intp
+        long n_pop_long, n_samples_long
+
+    # Note: the legacy rng will use `long` internally in either case
+    # If NumPy 2.0 is guaranteed at compile time this could also be written
+    # as `cnp.NPY_DEFAULT_INT == cnp.NPY_INTP`.
+    if np.int_ is np.intp:
+        # converting to long will allow conversion to integer (from float)
+        # via `int()`, but conversion from other objects does not.
```

Comment:
A little bit of rewording from reading the code.

```suggestion
        # converting to long via `int()` will allow conversion to
        # `intp`, but conversion from other objects does not.
```

Is this consistent with what you intent to say?

Reply:
Tried to tweak this and the other comments, maybe it helps.  The point is the second branch does `int()` implicitly and is the old behavior.  Cython doesn't do it implicitly in the first, so we do it explicitly.

Reply:
Turns out, this was a lot more confusing than I thought.  Because the definitions changed in the Cython 3 definitions file, which was the *actual* cause for changing how the conversion happened.

I am changing it back in NumPy 1.26.2 (and a slightly more thorough fix for 2.0).  So this will not be strictly necessary in the future (as the comment says).

(I actually prefer the strict version, but that is a separate discussion/transition.)

In either case, I hope the comment is relatively clear now.